### ISSUE TITLE
MaskZero() does not work for sequence-to-sequence learning?

### ISSUE THREAD

**Issue by cjmcmurtrie** [Element-Research/rnn#146]
I've raised this issue elsewhere but seems appropriate to create a new issue at this point. I don't believe that the :maskZero() function that decorates AbstractRecurrent.recurrentModule works with sequence to sequence models trained with variable length sequences and zero padding. This is because the hidden state is copied over from the encoder to the decoder, but reset when the first steps in the decoder are zeros.

A minimal code example is below (well, minimal - it is actually 162 lines of code!). It is basically a re-worked version of the example, but with the toy sequences padded with zeros, and the various maskZero layers/functions used. I have tried to write it in the style of the encoder-decoder example provided in the library, with one difference that here it is written to use the `optim` package. The output() function uses the model to reconstruct a sequence. If the training/test sequences are changed such that they do not contain zeros, this model works as normal.

Personally, for the reasons I gave, I don't think the current iteration of maskZero() can work for zero-padded sequence-to-sequence data. I think that rather than resetting states upon encountering zeros, they should be carried forward. Any help appreciated.
```lua
require 'nn'
require 'rnn'
require 'optim'
require 'cutorch'
require 'cunn'


local vector_dim = 20
local hidden = 1000
local vocab_size = 100
local max_len = 6


torch.manualSeed(123)

local model = {}

model.encoderLSTM = nn.FastLSTM(vector_dim, hidden):maskZero(1)
model.decoderLSTM = nn.FastLSTM(vector_dim, hidden):maskZero(1)

model.encoder = nn.Sequential()
					:add(nn.LookupTableMaskZero(vocab_size, vector_dim))
					:add(nn.SplitTable(1, 2))
					:add(nn.Sequencer(model.encoderLSTM))
					:add(nn.SelectTable(-1)):cuda()
model.encoder:zeroGradParameters()

model.decoder = nn.Sequential()
					:add(nn.LookupTableMaskZero(vocab_size, vector_dim))
					:add(nn.SplitTable(1, 2))
					:add(nn.Sequencer(model.decoderLSTM))
					:add(nn.Sequencer(nn.Linear(hidden, vocab_size)))
					:add(nn.Sequencer(nn.LogSoftMax())):cuda()
model.decoder:zeroGradParameters()

model.zerotensor = torch.Tensor(2):zero():cuda()
model.optimstate = {learningRate = 0.005, alpha = 0.95}
model.criterion = nn.SequencerCriterion(nn.MaskZeroCriterion(nn.ClassNLLCriterion(), 1)):cuda()

local encoder_input = torch.CudaTensor{{0, 0, 39, 40, 62, 63}, {0, 0, 0, 0, 98, 97}}
local decoder_input = torch.CudaTensor{{0, 1, 39, 40, 62, 63}, {0, 0, 0, 1, 98, 97}}
local decoder_target = torch.CudaTensor{{0, 39, 40, 62, 63, 100}, {0, 0, 0, 98, 97, 100}}


function forwardConnect(model, seq_len)

  model.decoderLSTM.userPrevOutput = 
  	nn.rnn.recursiveCopy(model.decoderLSTM.userPrevOutput, model.encoderLSTM.outputs[seq_len])
  model.decoderLSTM.userPrevCell = 
  	nn.rnn.recursiveCopy(model.decoderLSTM.userPrevCell, model.encoderLSTM.cells[seq_len])

end

function backwardConnect(model)

  model.encoderLSTM.userNextGradCell = 
  	nn.rnn.recursiveCopy(model.encoderLSTM.userNextGradCell, model.decoderLSTM.userGradPrevCell)
  model.encoderLSTM.gradPrevOutput = 
  	nn.rnn.recursiveCopy(model.encoderLSTM.gradPrevOutput, model.decoderLSTM.userGradPrevOutput)

end


-- Return a prediction for an input sequence
function output(input, model, max_out)

	model.encoderLSTM:remember()
	model.decoderLSTM:remember()

	model.encoder:forward(input)
	forwardConnect(model, input:size(1))

	local sequence = {}

	-- -- Forward <go> and all of it's output recursively back to the decoder
	local output = torch.CudaTensor{1}

	for i = 1, max_out do

		local prediction = model.decoder:forward(output)[1]

		-- prediction contains the probabilities for each word IDs.
		-- The index of the probability is the word ID.
		local prob, preds = prediction:sort(1, true)

		-- First one is the most likely.
		output = torch.CudaTensor{preds[1]}

		-- Terminate on EOS token
		if output[1] == 100 then
			break
		end

		table.insert(sequence, preds[1])
	end 

	model.encoder:forget()
	model.decoder:forget()

	return torch.Tensor(sequence)

end


function feval(x)

	if x ~= params then
        params:copy(x)
    end
    gradParams:zero()

    decoder_target_split = nn.SplitTable(1, 1):forward(decoder_target)

    -- Forward pass
    model.encoder:forward(encoder_input)
    forwardConnect(model, max_len)
    local decoder_output = model.decoder:forward(decoder_input)
    local loss = model.criterion:forward(decoder_output, decoder_target_split)

    -- Backward pass
    gEdec = model.criterion:backward(decoder_output, decoder_target_split)
	model.decoder:backward(decoder_input, gEdec)
	backwardConnect(model)
	model.encoder:backward(encoder_input, model.zerotensor)

	model.encoder:gradParamClip(2)
	model.decoder:gradParamClip(2)

	return loss, gradParams

end


modules = nn.Parallel()
    :add(model.encoder)
    :add(model.decoder)

params, gradParams = modules:getParameters()

local epochs = 1000

for e = 1 , epochs do
	_ , loss = optim.adam(feval, params, model.optimstate)
	print ('Loss:' .. tostring(loss[1]))
	if e % 100 == 0 then
		print (output(torch.CudaTensor{39, 40, 62}, model, 10))
	end
end
```

**Comment by cjmcmurtrie** [comment]
Just to update, I have found 50% of the solution to this. The other 50% I'm still working on.

Padding the decoder's data to the right, rather than the left, solves the problem of resetting states. So the dummy data above would be like this:
```
local encoder_input = torch.CudaTensor{{0, 0, 39, 40, 62, 63}, {0, 0, 0, 0, 98, 97}}
local decoder_input = torch.CudaTensor{{1, 39, 40, 62, 63, 0}, {1, 98, 97, 0, 0, 0}}
local decoder_target = torch.CudaTensor{{39, 40, 62, 63, 100, 0}, {98, 97, 100, 0, 0, 0}}
```
This seems to be the technically correct solution. However, there is still something strange going on with the gradient updates, where the final token seems to be getting more weight. Am still investigating.
